src/simulation_strategies/GeneralizedCliffordsSimulationStrategy.py

```diff
@@ -18,98 +17,108 @@ class GeneralizedCliffordsSimulationStrategy(SimulationStrategy):
     def __init__(self, interferometer_matrix: ndarray) -> None:
         self.r_sample = []
         self.number_of_input_photons = 0
-        self.pmfs = []  # Probability mass functions calculated along the way.
+        self.pmfs = dict()  # Probability mass functions calculated along the way. Keys should be ids of current r.
         self.interferometer_matrix = interferometer_matrix
-        self.input_state = []
-        self._labeled_states = defaultdict()
-        self.current_outputs = []
+        self.input_state = array([])
+        self._labeled_states = defaultdict(list)
+        self.possible_outputs = dict()
+        self.current_key = hash(tuple(self.r_sample))
```

Comment:
You can use the `tuple` as the key. Internally `hash` will be used and collisions problem will be solved.

Reply:
Right, thanks, it indeed seems to work in this case.